src/java.base/share/native/libzip/zip_util.c

```diff
@@ -213,45 +207,59 @@ InitializeZip()
     return 0;
 }
 
-/*
- * Reads len bytes of data into buf.
- * Returns 0 if all bytes could be read, otherwise returns -1.
- */
-static int
-readFully(ZFILE zfd, void *buf, jlong len) {
-  char *bp = (char *) buf;
-
-  while (len > 0) {
-        jlong limit = ((((jlong) 1) << 31) - 1);
-        jint count = (len < limit) ?
-            (jint) len :
-            (jint) limit;
-        jint n = ZFILE_read(zfd, bp, count);
-        if (n > 0) {
-            bp += n;
-            len -= n;
-        } else if (n == -1 && errno == EINTR) {
-          /* Retry after EINTR (interrupted by signal). */
-            continue;
-        } else { /* EOF or IO error */
-            return -1;
-        }
+static jlong readFullyAt(ZFILE zfd, void *buf, jlong nbytes, jlong offset) {
+  assert(nbytes >= 0);
+  assert(offset >= 0);
+#ifdef WIN32
+  jlong total = 0;
+  OVERLAPPED overlapped;
+  DWORD number_of_bytes_to_read;
+  DWORD number_of_bytes_read;
+  while (total < nbytes) {
+    memset(&overlapped, '\0', sizeof(overlapped));
+    overlapped.Offset = (DWORD) offset;
+    overlapped.OffsetHigh = (DWORD) (offset >> 32);
+    if ((nbytes - total) > 0xffffffffu) {
+      number_of_bytes_to_read = 0xffffffffu;
+    } else {
+      number_of_bytes_to_read = (DWORD) (nbytes - total);
     }
-    return 0;
-}
-
-/*
- * Reads len bytes of data from the specified offset into buf.
- * Returns 0 if all bytes could be read, otherwise returns -1.
- */
-static int
-readFullyAt(ZFILE zfd, void *buf, jlong len, jlong offset)
-{
-    if (IO_Lseek(zfd, offset, SEEK_SET) == -1) {
-        return -1; /* lseek failure. */
+    number_of_bytes_read = 0;
```

Comment:
do we really need to set number_of_bytes_read = 0 in every iteration ? .  As per MSDN it looks like ReadFile will do it implicitly.

[out, optional] lpNumberOfBytesRead

A pointer to the variable that receives the number of bytes read when using a synchronous hFile parameter. ReadFile sets this value to zero before doing any work or error checking. Use NULL for this parameter if this is an asynchronous operation to avoid potentially erroneous results.